File: app/macos/NativeStream/NativeStream/Core/M3UParser.swift
// NS-021 + NS-023: M3UParser
// Pure Swift M3U/M3U8 playlist parser. No third-party dependencies.
// Supports local file:// and remote https:// sources.

import Foundation

// MARK: - Parser warnings (non-fatal)

struct M3UParseWarning: Sendable {
    let line: Int
    let reason: String
}

// MARK: - Parser

actor M3UParser {

    // MARK: - Public API

    /// Parse from raw Data (used in unit tests without network).
    func parse(data: Data) throws -> (channels: [Channel], warnings: [M3UParseWarning]) {
        guard let text = String(data: data, encoding: .utf8) ??
                         String(data: data, encoding: .isoLatin1) else {
            throw AppError.playlistParseError(line: 0, reason: "File is not valid UTF-8 or Latin-1 text")
        }
        return parseText(text)
    }

    /// Parse from a local file:// or remote https:// URL.
    func parse(url: URL) async throws -> (channels: [Channel], warnings: [M3UParseWarning]) {
        let data: Data
        do {
            if url.isFileURL {
                data = try Data(contentsOf: url)
            } else {
                let (fetched, response) = try await URLSession.shared.data(from: url)
                guard let http = response as? HTTPURLResponse,
                      (200...299).contains(http.statusCode) else {
                    throw AppError.playlistFetchFailed(
                        url: url,
                        underlying: URLError(.badServerResponse)
                    )
                }
                data = fetched
            }
        } catch let error as AppError {
            throw error
        } catch {
            throw AppError.playlistFetchFailed(url: url, underlying: error)
        }
        return try parse(data: data)
    }

    // MARK: - Internal parsing

    private func parseText(_ text: String) -> (channels: [Channel], warnings: [M3UParseWarning]) {
        var channels: [Channel] = []
        var warnings: [M3UParseWarning] = []
        let lines = text.components(separatedBy: .newlines)

        var lineIndex = 0
        var pendingMeta: ChannelMetadata? = nil

        for (i, rawLine) in lines.enumerated() {
            lineIndex = i + 1
            let line = rawLine.trimmingCharacters(in: .whitespaces)

            if line.isEmpty { continue }

            if i == 0 && line.hasPrefix("#EXTM3U") {
                // Valid header — continue
                continue
            }

            if line.hasPrefix("#EXTINF:") {
                pendingMeta = parseExtInf(line, lineNumber: lineIndex, warnings: &warnings)
                continue
            }

            if line.hasPrefix("#") {
                // Other comment/directive — skip
                continue
            }

            // This should be a stream URL
            guard let streamURL = URL(string: line), streamURL.scheme != nil else {
                warnings.append(M3UParseWarning(line: lineIndex, reason: "Invalid stream URL: \(line)"))
                pendingMeta = nil
                continue
            }

            let meta = pendingMeta ?? ChannelMetadata(name: "Channel \(channels.count + 1)")
            channels.append(Channel(
                tvgId: meta.tvgId,
                name: meta.name,
                groupTitle: meta.groupTitle,
                logoURL: meta.logoURL,
                streamURL: streamURL
            ))
            pendingMeta = nil
        }

        return (channels, warnings)
    }

    // MARK: - EXTINF parsing

    private struct ChannelMetadata {
        var tvgId: String = ""
        var name: String
        var groupTitle: String = "Uncategorised"
        var logoURL: URL? = nil
    }

    private func parseExtInf(
        _ line: String,
        lineNumber: Int,
        warnings: inout [M3UParseWarning]
    ) -> ChannelMetadata? {
        // Format: #EXTINF:-1 tvg-id="ID" tvg-logo="URL" group-title="Group",Channel Name
        // The display name is after the last comma

        guard let commaRange = line.range(of: ",", options: .backwards) else {
            warnings.append(M3UParseWarning(line: lineNumber, reason: "EXTINF missing comma separator"))
            return nil
        }

        let name = String(line[commaRange.upperBound...]).trimmingCharacters(in: .whitespaces)
        let attrSection = String(line[line.startIndex..<commaRange.lowerBound])

        var meta = ChannelMetadata(name: name.isEmpty ? "Unknown Channel" : name)

        // Extract tvg-id
        if let value = extractAttribute("tvg-id", from: attrSection) {
            meta.tvgId = value
        }

        // Extract group-title
        if let value = extractAttribute("group-title", from: attrSection) {
            meta.groupTitle = value.isEmpty ? "Uncategorised" : value
        }

        // Extract tvg-logo
        if let value = extractAttribute("tvg-logo", from: attrSection),
           let logoURL = URL(string: value) {
            meta.logoURL = logoURL
        }

        return meta
    }

    /// Extracts a quoted attribute value from EXTINF attribute string.
    /// Handles both double-quoted and unquoted values.
    private func extractAttribute(_ key: String, from string: String) -> String? {
        // Match key="value" or key='value'
        let patterns = [
            "\(key)=\"([^\"]*)\"",
            "\(key)='([^']*)'",
        ]
        for pattern in patterns {
            if let regex = try? NSRegularExpression(pattern: pattern),
               let match = regex.firstMatch(in: string, range: NSRange(string.startIndex..., in: string)),
               let range = Range(match.range(at: 1), in: string) {
                return String(string[range])
            }
        }
        return nil
    }
}

File: app/android/app/src/main/java/com/nativestream/android/data/parser/M3uParser.kt
// app/src/main/java/com/nativestream/android/data/parser/M3uParser.kt
//
// M3U Parser
// Pure Kotlin M3U/M3U8 playlist parser. No third-party dependencies.
//   - Parses InputStream line-by-line — no full-file buffering
//   - Extracts tvg-id, tvg-logo, group-title from #EXTINF
//   - Skips malformed entries with a logged warning (no crash)
//   - Channels with missing tvg-id get empty string
//   - Detects url-tvg on the #EXTM3U header line
//
// Target: 10,000-channel M3U in < 500ms (verified by M3uParserTest)

package com.nativestream.android.data.parser

import android.util.Log
import com.nativestream.android.domain.model.Channel
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "M3uParser"

private const val HEADER_PREFIX    = "#EXTM3U"
private const val EXTINF_PREFIX    = "#EXTINF:"
private const val COMMENT_PREFIX   = "#"
private const val FALLBACK_GROUP   = "Uncategorised"

// Attribute keys used in #EXTINF lines
private const val ATTR_TVG_ID      = "tvg-id"
private const val ATTR_TVG_LOGO    = "tvg-logo"
private const val ATTR_GROUP_TITLE = "group-title"
private const val ATTR_URL_TVG     = "url-tvg"

// Regex patterns for key="value" and key='value' attribute extraction
private val DOUBLE_QUOTE_PATTERN = Regex("""(\S+)="([^"]*)"""")
private val SINGLE_QUOTE_PATTERN = Regex("""(\S+)='([^']*)'""")

private const val INITIAL_CHANNEL_CAPACITY  = 512
private const val INITIAL_WARNING_CAPACITY  = 16

@Singleton
class M3uParser @Inject constructor() {

    companion object {
        /** Extract only the EPG url-tvg from the M3U header — cheap, no full parse. */
        fun extractEpgUrl(bytes: ByteArray): String? {
            val text = bytes.toString(Charsets.UTF_8).let {
                if (it.contains('\uFFFD')) bytes.toString(Charsets.ISO_8859_1) else it
            }
            val firstLine = text.lineSequence().firstOrNull { it.isNotBlank() } ?: return null
            if (!firstLine.trimStart().startsWith(HEADER_PREFIX)) return null
            return DOUBLE_QUOTE_PATTERN.findAll(firstLine)
                .find { it.groupValues[1] == ATTR_URL_TVG }?.groupValues?.get(2)
                ?: SINGLE_QUOTE_PATTERN.findAll(firstLine)
                    .find { it.groupValues[1] == ATTR_URL_TVG }?.groupValues?.get(2)
        }
    }

    /**
     * Parse M3U content from a raw [ByteArray].
     * UTF-8 is attempted first; ISO-8859-1 (Latin-1) used as fallback — mirrors Swift behaviour.
     */
    fun parse(bytes: ByteArray): M3uParseResult {
        val text = bytes.toString(Charsets.UTF_8).let {
            // If the decoded text contains replacement characters, retry with Latin-1
            if (it.contains('\uFFFD')) bytes.toString(Charsets.ISO_8859_1) else it
        }
        return parse(text.byteInputStream())
    }

    /**
     * Parse M3U content from an [InputStream].
     * Reads line-by-line — no full-file buffering — safe for very large playlists.
     */
    fun parse(stream: InputStream): M3uParseResult {
        val channels  = ArrayList<Channel>(INITIAL_CHANNEL_CAPACITY)
        val warnings  = ArrayList<M3uParseWarning>(INITIAL_WARNING_CAPACITY)
        var epgUrl: String? = null
        var pendingMeta: ChannelMetadata? = null
        var lineNumber = 0

        BufferedReader(InputStreamReader(stream, Charsets.UTF_8)).use { reader ->
            reader.forEachLine { rawLine ->
                lineNumber++
                val line = rawLine.trim()
                if (line.isEmpty()) return@forEachLine

                when {
                    lineNumber == 1 && line.startsWith(HEADER_PREFIX) -> {
                        epgUrl = extractAttributeValue(ATTR_URL_TVG, from = line)
                    }

                    line.startsWith(EXTINF_PREFIX) -> {
                        pendingMeta = parseExtInf(line, lineNumber, warnings)
                    }

                    line.startsWith(COMMENT_PREFIX) -> {
                        // Other directive or comment — skip
                    }

                    else -> {
                        // Should be a stream URL
                        if (!looksLikeUrl(line)) {
                            warnings.add(M3uParseWarning(lineNumber, "Invalid stream URL: $line"))
                            Log.w(TAG, "Line $lineNumber — invalid stream URL, skipping")
                            pendingMeta = null
                            return@forEachLine
                        }

                        val meta = pendingMeta ?: ChannelMetadata(
                            name = "Channel ${channels.size + 1}",
                        )
                        channels.add(
                            Channel.create(
                                tvgId      = meta.tvgId,
                                name       = meta.name,
                                groupTitle = meta.groupTitle,
                                logoUrl    = meta.logoUrl,
                                streamUrl  = line,
                            )
                        )
                        pendingMeta = null
                    }
                }
            }
        }

        Log.d(TAG, "Parsed ${channels.size} channels, ${warnings.size} warnings")
        return M3uParseResult(channels = channels, epgUrl = epgUrl, warnings = warnings)
    }

    // ── EXTINF line parsing ───────────────────────────────────────────────────

    /**
     * Parses a single #EXTINF line into a [ChannelMetadata].
     * Format: #EXTINF:-1 tvg-id="…" tvg-logo="…" group-title="…",Display Name
     * Display name is everything after the last comma.
     */
    private fun parseExtInf(
        line: String,
        lineNumber: Int,
        warnings: MutableList<M3uParseWarning>,
    ): ChannelMetadata? {
        val commaIndex = line.lastIndexOf(',')
        if (commaIndex < 0) {
            warnings.add(M3uParseWarning(lineNumber, "EXTINF missing comma separator"))
            return null
        }

        val displayName  = line.substring(commaIndex + 1).trim()
        val attributeSection = line.substring(0, commaIndex)

        val name       = displayName.ifEmpty { "Unknown Channel" }
        val tvgId      = extractAttributeValue(ATTR_TVG_ID,      from = attributeSection) ?: ""
        val groupTitle = extractAttributeValue(ATTR_GROUP_TITLE,  from = attributeSection)
            ?.takeUnless { it.isEmpty() } ?: FALLBACK_GROUP
        val logoUrl    = extractAttributeValue(ATTR_TVG_LOGO,     from = attributeSection)

        return ChannelMetadata(
            tvgId      = tvgId,
            name       = name,
            groupTitle = groupTitle,
            logoUrl    = logoUrl,
        )
    }

    // ── Attribute extraction ──────────────────────────────────────────────────

    /**
     * Extracts a quoted attribute value from an EXTINF attribute string.
     * Handles both double-quoted and single-quoted values — mirrors Swift impl.
     */
    private fun extractAttributeValue(key: String, from: String): String? {
        // Try key="value" first, then key='value'
        DOUBLE_QUOTE_PATTERN.findAll(from).forEach { match ->
            if (match.groupValues[1] == key) return match.groupValues[2]
        }
        SINGLE_QUOTE_PATTERN.findAll(from).forEach { match ->
            if (match.groupValues[1] == key) return match.groupValues[2]
        }
        return null
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    /**
     * Lightweight URL sanity check — avoids constructing a full URI object for
     * every line. Mirrors Swift's `URL(string:)?.scheme != nil` guard.
     */
    private fun looksLikeUrl(line: String): Boolean =
        line.startsWith("http://")  ||
                line.startsWith("https://") ||
                line.startsWith("rtmp://")  ||
                line.startsWith("rtsp://")

    // ── Internal metadata accumulator ────────────────────────────────────────

    private data class ChannelMetadata(
        val tvgId:      String  = "",
        val name:       String,
        val groupTitle: String  = FALLBACK_GROUP,
        val logoUrl:    String? = null,
    )
}